adr/Registry Refactoring Design.md

```diff
@@ -0,0 +1,144 @@
+# Registry Refactoring Design
+
+### Introduction
+
+Currently the `Registry Client` in `go-mod-registry` module provides **Service Configuration** and **Service Registration** functionality. The goal of this design is to refactor the `go-mod-registry` module for separation of concerns. The **Service Registry** functionality will stay in the `go-mod-registry` module and the **Service Configuration** functionality will be separated out into a new `go-mod-configuration` module. This allows for implementations for deferent providers for each, another aspect of separation of concerns.
+
+### Provider Connection information
+
+An aspect of using the current `Registry Client` is "*Where do the services get the `Registry Provider` connection information?*" Currently all services either pull this connection information from the local configuration file or from the `edgex_registry` environment variable. Device Services also have the option to specify this connection information on the command line. With the refactoring for separation of concerns, this issue changes to "*Where do the services get the `Configuration Provider` connection information?*"
+
+There have been concerns voiced by some in the EdgeX community that storing this `Configuration Provider` connection information in the configuration which ultimately is provided by that provider is not the right design.
+
+This design proposes that all services will use the command line option approach with the ability to override with an environment variable. The  `Configuration Provider` information will **<u>not</u>** be stored in each service's local configuration file. The `edgex_registry` environment variable will be deprecated. The `Registry Provider` connection information will continue to be stored in each service's configuration either locally or from the`Configuration Provider` same as all other Edgex Client and Database connection information. An additional property will be added the the Registry Provider information to indicate if it is `Enabled`.
+
+### Command line option changes
+
+The new `-cp/-configProvider` command line option will be added to each service which will have a value specified using the format `{type}.{protocol}://{host}:{port}` e.g `consul.http://localhost:8500`.  This new command line option will be overridden by the `edgex_configuration_provider` environment variable when it is set. This environment variable's value has the same format as the command line option value. 
+
+If **no value** is provided to the `-cp/-configProvider` option, i.e. just `-cp`, and no environment variable override is specified, the default value of `consul.http://localhost:8500` will be used. 
+
+if `-cp/-configProvider` **not used** and no environment variable override is specified the local configuration file is used, as is it now.
+
```

Comment:
what happens if the local configuration file has no cp entry? use the `consul.http://localhost:8500`?

A little concerned if services start-up out of sync, that is one referring to a registry foo and another to bar.
Guess an error message may work  that attempts a connection check to registry and if it fails, emit an misconfiguration error message and promptly exit. 

A la dependency injection, would it make sense to